packages/react-aria-components/test/ListBox.test.js

```diff
@@ -527,6 +527,98 @@ describe('ListBox', () => {
     expect(document.activeElement).toBe(options[1]);
   });
 
+  it('should not throw TypeError at boundaries of grid layout', async () => {
+    /**
+     * The following ListBox is roughly in this shape:
+     * 
+     * -------------------
+     * | 1,1 | 1,2 | 1,3 |
+     * -------------------
+     * | 2,1 | 2,2 | 2,3 |
+     * -------------------
+     * | 3,1 | 3,2 | 3,3 |
+     * -------------------
+     */
+    let {getAllByRole, rerender} = render(
+      <ListBox layout="grid" aria-label="Test">
+        <ListBoxItem>1,1</ListBoxItem>
+        <ListBoxItem>1,2</ListBoxItem>
+        <ListBoxItem>1,3</ListBoxItem>
+        <ListBoxItem>2,1</ListBoxItem>
+        <ListBoxItem>2,2</ListBoxItem>
+        <ListBoxItem>2,3</ListBoxItem>
+        <ListBoxItem>3,1</ListBoxItem>
+        <ListBoxItem>3,2</ListBoxItem>
+        <ListBoxItem>3,3</ListBoxItem>
+      </ListBox>);
+    let options = getAllByRole('option');
+
+    jest.spyOn(HTMLElement.prototype, 'getBoundingClientRect').mockImplementation(function () {
+      if (this.getAttribute('role') === 'listbox') {
+        return {top: 0, left: 0, bottom: 200, right: 300, width: 300, height: 200};
+      } else {
+        let index = [...this.parentElement.children].indexOf(this);
+        return {top: Math.floor(index / 3) * 40, left: (index % 3) * 100, bottom: Math.floor(index / 3) * 40 + 40, right: (index % 3) * 100 + 100, width: 100, height: 40};
+      }
+    });
+
+    await user.tab();
+    expect(document.activeElement).toBe(options[0]);  // 1,1
+
+    keyPress('ArrowDown');
+    expect(document.activeElement).toBe(options[3]);  // 2,1
+
+    keyPress('ArrowDown');
+    expect(document.activeElement).toBe(options[6]);  // 3,1
+
+    keyPress('ArrowDown');  // shouldn't throw
+    expect(document.activeElement).toBe(options[6]);  // 3,1
+
+    keyPress('ArrowUp');
+    expect(document.activeElement).toBe(options[3]);  // 2,1
+
+    keyPress('ArrowUp');
+    expect(document.activeElement).toBe(options[0]);  // 1,1
+
+    keyPress('ArrowUp');  // shouldn't throw
+    expect(document.activeElement).toBe(options[0]);  // 1,1
+
+    // horizontal
+    rerender(
```

Comment:
do we not need to adjust the positions again? might be easier to have two tests?

Reply:
Sorry, I think it isn't very clear what this test is doing.

I was basically traversing the 3x3 grid down and back up, along the first column (for vertical orientation), and then going right and coming back left, along the first row (for horizontal orientation). See the below video for visualization.


https://github.com/adobe/react-spectrum/assets/71210554/d462a55d-aa94-45f4-81b4-25e2e972528e

And I think you're right, this test is unnecessarily confusing haha. I'll go ahead and split the test in two for better readability.

Reply:
apologies if my head just won't wrap around this, but shouldn't the collection go from
```
horizontal
     * -------------------
     * | 1,1 | 1,2 | 1,3 |
     * -------------------
     * | 2,1 | 2,2 | 2,3 |
     * -------------------
     * | 3,1 | 3,2 | 3,3 |
     * -------------------
 
to
vertical
     * -------------------
     * | 1,1 | 2,1 | 3,1 |
     * -------------------
     * | 1,2 | 2,2 | 3,2 |
     * -------------------
     * | 1,3 | 2,3 | 3,3 |
     * -------------------
```
otherwise the

Reply:
Yeah you're right sorry about that! I wasn't paying attention to the order of the items inside the grid for a horizontal grid at all. I fixed it with the below commit, though the numbers are arranged opposite to your diagram. 

I hope this behavior in the video below is what you wanted to see?


https://github.com/adobe/react-spectrum/assets/71210554/cfbdd130-72f6-4be3-825d-09230478faf7

Reply:
what software did you use to record that video, @sookmax?

Reply:
@staticshock I use [Mac's Screenshot](https://support.apple.com/guide/mac-help/take-a-screenshot-mh26782/mac) and [KeyCastr](https://github.com/keycastr/keycastr). You probably asked for KeyCastr. Give it a try, it's awesome!